File: finchley/reservation-client/src/main/java/com/example/reservationclient/ReservationClientApplication.java
package com.example.reservationclient;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.rsocket.Payload;
import io.rsocket.RSocketFactory;
import io.rsocket.transport.netty.client.TcpClientTransport;
import io.rsocket.util.DefaultPayload;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.reactivestreams.Publisher;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.gateway.filter.ratelimit.RedisRateLimiter;
import org.springframework.cloud.gateway.route.RouteLocator;
import org.springframework.cloud.gateway.route.builder.RouteLocatorBuilder;
import org.springframework.cloud.netflix.hystrix.HystrixCommands;
import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpHeaders;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;

import java.io.IOException;

import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@SpringBootApplication
public class ReservationClientApplication {

	public static void main(String[] args) {
		SpringApplication.run(ReservationClientApplication.class, args);
	}


	@Bean
	MapReactiveUserDetailsService authentication() {
		// @rob_winch became sad :-(
		UserDetails jlong = User.withDefaultPasswordEncoder().username("jlong").password("pw").roles("USER").build();
		UserDetails rwinch = User.withDefaultPasswordEncoder().username("rwinch").password("pw").roles("USER", "ADMIN").build();
		return new MapReactiveUserDetailsService(jlong, rwinch);
	}

	@Bean
	SecurityWebFilterChain authorization(ServerHttpSecurity http) {
		http.httpBasic();
		http.authorizeExchange()
			.pathMatchers("/proxy").authenticated()
			.anyExchange().permitAll();
		http.csrf().disable();
		return http.build();
	}

	@Bean
	RedisRateLimiter redisRateLimiter() {
		return new RedisRateLimiter(5, 7);
	}


	@Bean
	RouterFunction<ServerResponse> adapter(ReservationClient client) {
		return route(GET("/reservations/names"),
			serverRequest -> {

				Flux<String> names = client
					.getAllReservations()
					.map(Reservation::getReservationName);

				Publisher<String> fallback = HystrixCommands
					.from(names)
					.fallback(Flux.just("EEK!"))
					.commandName("names")
					.eager()
					.build();

				return ServerResponse.ok().body(fallback, String.class);
			});
	}

	@Bean
	RouteLocator gateway(RouteLocatorBuilder rlb) {
		return rlb
			.routes()
			.route(
				rSpec ->
					rSpec
						.host("*.foo.ca").and().path("/proxy")
						.filters(fSpec -> fSpec
							.setPath("/reservations")
							.addResponseHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
							.requestRateLimiter(rc -> rc.setRateLimiter(redisRateLimiter()))
						)
						.uri("http://localhost:8080")
			)
			.build();
	}

	@Bean
	WebClient client(WebClient.Builder builder) {
		return builder.build();
	}

}


@Data
@AllArgsConstructor
@NoArgsConstructor
class Reservation {
	private Integer id;
	private String reservationName;
}
/*

@Component
class ReservationClient {

	private final WebClient webClient;

	ReservationClient(WebClient webClient) {
		this.webClient = webClient;
	}

	Flux<Reservation> getAllReservations() {
		return this.webClient
			.get()
			.uri("http://localhost:8080/reservations")
			.retrieve()
			.bodyToFlux(Reservation.class);
	}
}
*/


@Component
class ReservationClient {


	private final ObjectMapper objectMapper;

	ReservationClient(ObjectMapper objectMapper) {
		this.objectMapper = objectMapper;
	}

	Reservation from(String json) {
		try {
			return this.objectMapper.readValue(json, Reservation.class);
		}
		catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	Flux<Reservation> getAllReservations() {
		return RSocketFactory
			.connect()
			.transport(TcpClientTransport.create(7000))
			.start()
			.flatMapMany(rs -> rs.requestStream(DefaultPayload.create(new byte[0])))
			.map(Payload::getDataUtf8)
			.map(this::from);

	}
}

File: finchley/kotlin/reservation-client/src/main/kotlin/com/example/reservationclient/ReservationClientApplication.kt
package com.example.reservationclient

import org.reactivestreams.Publisher
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.cloud.client.loadbalancer.reactive.LoadBalancerExchangeFilterFunction
import org.springframework.cloud.gateway.filter.ratelimit.RedisRateLimiter
import org.springframework.cloud.gateway.route.builder.RouteLocatorBuilder
import org.springframework.cloud.gateway.route.builder.filters
import org.springframework.cloud.gateway.route.builder.routes
import org.springframework.cloud.netflix.hystrix.HystrixCommands
import org.springframework.cloud.stream.annotation.EnableBinding
import org.springframework.cloud.stream.messaging.Source
import org.springframework.context.annotation.Bean
import org.springframework.messaging.support.MessageBuilder
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService
import org.springframework.security.core.userdetails.User
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToFlux
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.bodyToMono
import org.springframework.web.reactive.function.server.router
import reactor.core.publisher.Flux

@EnableBinding(Source::class)
@SpringBootApplication
class ReservationClientApplication() {

	@Bean
	fun client(eff: LoadBalancerExchangeFilterFunction) = WebClient.builder().filter(eff).build()

	@Bean
	fun routes(client: WebClient, source: Source) = router {
		POST("/reservations") { request ->
			val send: Publisher<Boolean> = request.bodyToMono<Reservation>()
					.map { it.reservationName }
					.map { MessageBuilder.withPayload(it).build() }
					.map { source.output().send(it) }
			ServerResponse.ok().body(send)
		}
		GET("/reservations/names") {
			val names: Publisher<String> = client
					.get()
					.uri("http://reservation-service/reservations")
					.retrieve()
					.bodyToFlux<Reservation>()
					.map { it.reservationName }
			val circuit: Publisher<String> = HystrixCommands
					.from(names)
					.commandName("reservation-names")
					.eager()
					.fallback(Flux.just("EEK!"))
					.build()
			ServerResponse.ok().body(circuit)
		}
	}

	@Bean
	fun authentication() =
			MapReactiveUserDetailsService(User.withDefaultPasswordEncoder()
					.username("user")
					.password("password")
					.roles("USER")
					.build())

	@Bean
	fun authorization(httpSecurity: ServerHttpSecurity) =
			httpSecurity
					.authorizeExchange().pathMatchers("/rl").authenticated()
					.anyExchange().permitAll()
					.and()
					.httpBasic()
					.and()
					.csrf().disable()
					.build()

	@Bean
	fun rateLimiter() = RedisRateLimiter(4, 6)

	@Bean
	fun gateway(rlb: RouteLocatorBuilder, rl: RedisRateLimiter) =
			rlb.routes {
				route {
					path("/rl")
					filters {
						requestRateLimiter({ it.rateLimiter = rl })
						setPath("/reservations")
					}
					uri("lb://reservation-service")
				}
				route {
					path("/lb")
					filters {
						setPath("/reservations")
					}
					uri("lb://reservation-service")
				}
			}
}

fun main(args: Array<String>) {
	runApplication<ReservationClientApplication>(*args)
}


class Reservation(
		val id: String? = null,
		val reservationName: String? = null)